From f6a426c58dfb38a0e23ed6690d18b4f7bd73cb1c Mon Sep 17 00:00:00 2001
From: Frans Klaver <frans.klaver@vislink.com>
Date: Tue, 24 Jun 2025 11:22:04 +0200
Subject: [PATCH] driver: apdu: mbim: sleep while waiting for sim

When waiting for the sim to become available after potentially
switching, we don't actually do much waiting. We're basically only
delayed by the speed of the communication here. If the switch doesn't
happen quickly enough, this may result in us deciding a sim didn't
become available, but subsequent calls will produce usable results.

So introduce a small 50 ms wait time, so we limit the total waiting
time to about 1 second. This I think should be enough for the sim to
actually become available.

resolve #241

Signed-off-by: Frans Klaver <frans.klaver@vislink.com>
---
 driver/apdu/mbim.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/driver/apdu/mbim.c b/driver/apdu/mbim.c
index 75fe84e0..ef9f5889 100644
--- a/driver/apdu/mbim.c
+++ b/driver/apdu/mbim.c
@@ -107,6 +107,11 @@ static int select_sim_slot(struct mbim_data *mbim_priv)
         if (is_sim_available(mbim_priv)) {
             return 0;
         }
+        struct timespec ts = {
+            .tv_sec = 0,
+            .tv_nsec = 50000000
+        };
+        nanosleep(&ts, NULL);
     }
 
     fprintf(stderr, "sim did not become available\n");
